File: packages/webview_flutter/webview_flutter_android/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterViewFactory.java
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.webviewflutter;

import android.content.Context;
import android.view.View;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import io.flutter.plugin.common.StandardMessageCodec;
import io.flutter.plugin.platform.PlatformView;
import io.flutter.plugin.platform.PlatformViewFactory;

class FlutterViewFactory extends PlatformViewFactory {
  private final AndroidWebkitLibraryPigeonInstanceManager instanceManager;

  FlutterViewFactory(AndroidWebkitLibraryPigeonInstanceManager instanceManager) {
    super(StandardMessageCodec.INSTANCE);
    this.instanceManager = instanceManager;
  }

  @NonNull
  @Override
  public PlatformView create(Context context, int viewId, @Nullable Object args) {
    final Integer identifier = (Integer) args;
    if (identifier == null) {
      throw new IllegalStateException("An identifier is required to retrieve a View instance.");
    }

    final Object instance = instanceManager.getInstance(identifier);

    if (instance instanceof PlatformView) {
      return (PlatformView) instance;
    } else if (instance instanceof View) {
      return new PlatformView() {
        @Override
        public View getView() {
          return (View) instance;
        }

        @Override
        public void dispose() {}
      };
    }

    throw new IllegalStateException(
        "Unable to find a PlatformView or View instance: " + args + ", " + instance);
  }
}

File: packages/webview_flutter/webview_flutter_wkwebview/darwin/webview_flutter_wkwebview/Sources/webview_flutter_wkwebview/FlutterViewFactory.swift
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import Foundation

#if os(iOS)
  import Flutter
  import UIKit
#elseif os(macOS)
  import FlutterMacOS
#else
  #error("Unsupported platform.")
#endif

#if os(iOS)
  class PlatformViewImpl: NSObject, FlutterPlatformView {
    // TODO(bparrishMines): Change to strong reference once this issue is fixed in the engine and
    // makes it to stable. See https://github.com/flutter/flutter/issues/168535.
    // The InstanceManager used by pigeon adds an associated object to added instances that makes a message call when
    // they are deallocated. This sets a weak reference to the underlying UIView to prevent a crash where the UIView is
    // no longer referenced by the plugin, but the FlutterViewController still maintains a transitive reference to it
    // when the BinaryMessenger becomes invalid.
    weak var uiView: UIView?

    init(uiView: UIView) {
      self.uiView = uiView
    }

    func view() -> UIView {
      if let uiView = uiView {
        return uiView
      }

      NSLog(
        "WebViewFlutterPluginError: UIView has been deallocated, but is still being requested as a PlatformView."
      )
      return UIView()
    }
  }
#endif

/// Implementation of `FlutterPlatformViewFactory` that retrieves the view from the `WebKitLibraryPigeonInstanceManager`.
class FlutterViewFactory: NSObject, FlutterPlatformViewFactory {
  unowned let instanceManager: WebKitLibraryPigeonInstanceManager

  init(instanceManager: WebKitLibraryPigeonInstanceManager) {
    self.instanceManager = instanceManager
  }

  #if os(iOS)
    func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?)
      -> FlutterPlatformView
    {
      let identifier: Int64 = args is Int64 ? args as! Int64 : Int64(args as! Int32)
      let instance: AnyObject? = instanceManager.instance(forIdentifier: identifier)

      let view = instance as! UIView
      view.frame = frame
      return PlatformViewImpl(uiView: view)
    }
  #elseif os(macOS)
    func create(
      withViewIdentifier viewId: Int64,
      arguments args: Any?
    ) -> NSView {
      let identifier: Int64 = args is Int64 ? args as! Int64 : Int64(args as! Int32)
      let instance: AnyObject? = instanceManager.instance(forIdentifier: identifier)
      return instance as! NSView
    }
  #endif

  #if os(iOS)
    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
      return FlutterStandardMessageCodec.sharedInstance()
    }
  #elseif os(macOS)
    func createArgsCodec() -> (FlutterMessageCodec & NSObjectProtocol)? {
      return FlutterStandardMessageCodec.sharedInstance()
    }
  #endif
}
